### ISSUE TITLE
"restler_custom_payload..." parameter options in dict.json

### ISSUE THREAD

**Issue by mbiuki** [microsoft/restler-fuzzer#353]
I have made some confusion to myself about "`restler_custom_payload_`" parameters that are available for `dict.json`.

Could you please explain what is the difference between `restler_custom_payload` and `restler_custom_payload_query`?

The query parameters are same as body parameters, are they not?

Also their difference with:
`restler_custom_payload_uuid4_suffix`?

**Comment by PatGod** [comment]
`restler_custom_payload `are for request body parameters.

`restler_custom_payload_query` are similar but for the query part of REST API requests.

`restler_custom_payload_uuid4_suffix` can be used to specify a constant value to which will be appended a randomly-generated suffix.

**Comment by marina-p** [comment]
Hello @mbiuki,

```restler_custom_payload_query``` - will only be used for query parameter substitution (```PUT /api?query_param1=x&query_param2=y```).  If the parameter is not declared in the spec, it will be injected.

 ```restler_custom_payload_header``` - will only be used for request header parameter substitution.  If the parameter is not declared in the spec, it will be injected.

```restler_custom_payload``` - will be used to substitute any parameter or property with that name, whether in the path, query, header, or body.  The difference from the above two others is that this is only a replacement, no extra parameters can be injected via this construct.  

As you can see, there is no ```restler_custom_payload_path``` or ```restler_custom_payload_body``` where the replacement would have to be made only in the path or only in the body.  (Not present because we have not seen these use cases yet.)

Thanks,

Marina

**Comment by mbiuki** [comment]
- So if I have path parameters like `/agent/{agentId}/settings/{agentKey}`, you'd suggest that I add their UUID in `restler_custom_payload`?
- How does the injection work? Does it count toward `max_combinations`?

**Comment by marina-p** [comment]
1. It depends.  If 'agentId' or 'agentKey' are constants that must be specified by the user, and the 'agent' and 'key' are not created by the API, then yes ```restler_custom_payload``` is needed to assign values to them.  If these are created by the API (e.g. with PUT /agent), then there is no need to specify the constant since RESTler can create a new one and assign these via producer-consumer dependencies.

2. For all grammar elements that refer to the dictionary (custom payloads, fuzzables), if there is more than one value defined in the dictionary, then yes several combinations are tested and this "counts" toward ```max_combinations```.  e.g. if you test two different API versions by using ```restler_custom_payload_query("api-version")```, this will cause twice the number of combinations as if you only test the one version.

**Comment by mbiuki** [comment]
How about the header parameters:
```yaml
      parameters:
        - name: X-Work-ID
          in: header
          description: Work ID.
          required: true
          schema:
            type: string
            format: uuid
```

I think these would definitely need to be specified by `restler_custom_payload_header` as there could be no way by RESTLer to infer them?

**Comment by marina-p** [comment]
RESTler will soon support inferring dependencies for headers as well.  Is this work ID created by another PUT/POST in the API?

**Comment by mbiuki** [comment]
No, it must be given beforehand or taken from a different API (e.g., `GET https://api.test.com/getmyworkid/v2`). Although I am not sure how inference may be possible, this would be a great add-on to the fuzzing service and would help a lot with autonomy. 

The exact UUID of the `X-Work-ID` must be sent in the header of a GET/PUT/DELETE to have the request work o.w. the server will issue a 401 response.

https://github.com/microsoft/restler-fuzzer/issues/353#issuecomment-931670597 was very related to the problem I am facing as described in #348.

**Comment by marina-p** [comment]
Ok - yes, for now you should specify the custom payload for ```X-Work-ID``` in the dictionary and recompile to have the custom payload added to the grammar.

Dependencies w/ a completely different API, correct (w/ a different host are not supported yet, related issue #334 that was opened recently would enable this (this is not on our near-term road map).

**Comment by marina-p** [comment]
I'll take a closer look and reply separately on #348.
